packages/firestore/test/unit/local/web_storage_metadata_channel.test.ts

```diff
@@ -0,0 +1,271 @@
+ * limitations under the License.
+ */
+
+import * as persistenceHelpers from './persistence_test_helpers';
+import {
+  InstanceStateSchema,
+  WebStorageMetadataChannel,
+  InstanceMetadataChannel
+} from '../../../src/local/instance_metadata_channel';
+import { BatchId, TargetId } from '../../../src/core/types';
+import { AutoId } from '../../../src/util/misc';
+import { expect } from 'chai';
+
+const GRACE_INTERVAL_MS = 100;
+
+describe('WebStorageMetadataChannel', () => {
+  if (!WebStorageMetadataChannel.isAvailable()) {
+    console.warn(
+      'No LocalStorage. Skipping LocalStorageNotificationChannelTests tests.'
+    );
+    return;
+  }
+
+  const localStorage = window.localStorage;
+
+  let notificationChannel: InstanceMetadataChannel;
+  let ownerId;
+
+  beforeEach(() => {
+    ownerId = AutoId.newId();
+  });
+
+  function assertInstanceState(
+    activeTargetIds: number[],
+    minMutationBatchId?: number,
+    maxMutationBatchId?: number
+  ): void {
+    const actual = JSON.parse(
+      localStorage.getItem(
+        `fs_instances_${persistenceHelpers.TEST_PERSISTENCE_PREFIX}_${ownerId}`
+      )
+    );
+
+    let expectedKeyCount = /* lastUpdateTime + activeTargetIds */ 2;
+
+    expect(actual.lastUpdateTime).to.be.a('number');
+    expect(actual.lastUpdateTime).to.be.greaterThan(
+      Date.now() - GRACE_INTERVAL_MS
+    );
+    expect(actual.lastUpdateTime).to.be.at.most(Date.now());
+
+    expect(actual.activeTargetIds).to.be.an('array');
+    expect(actual.activeTargetIds).to.be.deep.equal(activeTargetIds);
+
+    if (minMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.minMutationBatchId).to.be.equal(minMutationBatchId);
+    }
+
+    if (maxMutationBatchId !== undefined) {
+      ++expectedKeyCount;
+      expect(actual.maxMutationBatchId).to.be.equal(maxMutationBatchId);
+    }
+
+    expect(Object.keys(actual).length).to.be.equal(expectedKeyCount);
+  }
+
+  describe('persists mutation batches', () => {
+    beforeEach(() => {
+      return persistenceHelpers
+        .testLocalStorageNotificationChannel(ownerId, [], [])
+        .then(nc => {
+          notificationChannel = nc;
+        });
+    });
+
+    afterEach(() => {
+      notificationChannel.shutdown();
+    });
+
+    it('with empty batch', () => {
+      assertInstanceState([]);
+    });
+
+    it('with one batch', () => {
+      notificationChannel.addLocalPendingMutation(0);
+      assertInstanceState([], 0, 0);
+    });
+
+    it('with multiple batches', () => {
+      notificationChannel.addLocalPendingMutation(0);
+      notificationChannel.addLocalPendingMutation(1);
+      assertInstanceState([], 0, 1);
+
+      notificationChannel.addLocalPendingMutation(2);
+      notificationChannel.addLocalPendingMutation(3);
+      assertInstanceState([], 0, 3);
+
+      notificationChannel.removeLocalPendingMutation(0);
+      notificationChannel.removeLocalPendingMutation(2);
```

Comment:
FWIW I think this is an invalid operation (mutation batches must always be removed in order). I was actually tempted to have you assert it in the remove implementation, but I think it's asserted elsewhere so it seemed overkill.

I'd be inclined to either change this to 1 or add a comment that this is technically invalid.

Reply:
I added a comment.